crypto/ml_kem/ml_kem-c.inc

```diff
@@ -0,0 +1,240 @@
+#include "ml_kem_local.h"
+#ifndef NDEBUG
+# include <stdio.h>
+#endif
+
+/*
+ * Variant-independence macros, variously shortened for convenience and
+ * internal consumption by original BoringSSL code.
+ */
+# if BITS == 512
+# define kemsym(a, b)       a##_512_##b
+# elif BITS == 768
+# define kemsym(a, b)       a##_768_##b
+# elif BITS == 1024
+# define kemsym(a, b)       a##_1024_##b
+# else
+#  error "Unsupported ML-KEM variant bit size"
+# endif
+
+/* Constants */
+# define DEGREE         ML_KEM_DEGREE
+# define RANK           kemsym(ML_KEM,RANK)
+# define ETA1           kemsym(ML_KEM,ETA1)
+# define ETA2           kemsym(ML_KEM,ETA2)
+# define DU             kemsym(ML_KEM,DU)
+# define DV             kemsym(ML_KEM,DV)
+# define SECBITS        kemsym(ML_KEM,RNGSEC)
+# define CTEXT_BYTES    ML_KEM_CTEXT_BYTES(RANK, DU, DV)
+# define PUBKEY_BYTES   ML_KEM_PUBKEY_BYTES(RANK)
+# define PRVKEY_BYTES   ML_KEM_PRVKEY_BYTES(RANK)
+# define VECTOR_BYTES   ML_KEM_VECTOR_BYTES(RANK)
+
+# if (DEGREE != 256) || (ML_KEM_PRIME != 3329) || (ML_KEM_LOG2PRIME != 12) || \
+     (DU < 2) || (DU >= 12) || (DV < 2) || (DV >= 12) || \
+     (ETA1 < 2) || (ETA1 > 3) || (ETA2 < 2) || (ETA2 > 3)
+#  error "Invalid ML-KEM variant-independent constants"
+# endif
+# define U_VECTOR_BYTES (DU * RANK * DEGREE / 8)
+
+# if ETA1 == 2
+#  define cbd_eta1      ossl_ml_kem_cbd_2
+# elif ETA1 == 3
+#  define cbd_eta1      ossl_ml_kem_cbd_3
+# endif
+# if ETA2 == 2
+#  define cbd_eta2      ossl_ml_kem_cbd_2
+# elif ETA2 == 3
+#  define cbd_eta2      ossl_ml_kem_cbd_3
+# endif
+
+/* Types */
+typedef kemsym(ossl_ml_kem,vector)      vector;
+typedef kemsym(ossl_ml_kem,public_key)  public_key;
+typedef kemsym(ossl_ml_kem,private_key) private_key;
+
+/* Main keypair gen/encap/decap API */
+# define genkey_rand                kemsym(ossl_ml_kem,genkey_rand)
+# define encap_rand                 kemsym(ossl_ml_kem,encap_rand)
+# define decap                      kemsym(ossl_ml_kem,decap)
+
+/* For testing and encoding rather than KEM computation */
+# define genkey_seed                kemsym(ossl_ml_kem,genkey_seed)
+# define encap_seed                 kemsym(ossl_ml_kem,encap_seed)
+# define parse_public_key           kemsym(ossl_ml_kem,parse_public_key)
+# define encode_public_key          kemsym(ossl_ml_kem,encode_public_key)
+# define parse_private_key          kemsym(ossl_ml_kem,parse_private_key)
+# define encode_private_key         kemsym(ossl_ml_kem,encode_private_key)
+# define get_vinfo                  kemsym(ossl_ml_kem,get_vinfo)
+
+typedef const ossl_ml_kem_vinfo *vinfo_t;
+
+/*
+ * Per-variant fixed parameters
+ */
+static const ossl_ml_kem_vinfo vinfo = {
+    cbd_eta1,
+    cbd_eta2,
+    VECTOR_BYTES,
+    PRVKEY_BYTES,
+    PUBKEY_BYTES,
+    CTEXT_BYTES,
+    U_VECTOR_BYTES,
+    sizeof(public_key),
+    sizeof(private_key),
+    BITS,
+    RANK,
+    DU,
+    DV,
+    SECBITS,
+};
+
+void encode_public_key(uint8_t *out, const public_key *pub)
```

Comment:
Needs to be static or start with `ossl_`.
Here and elsewhere.

Reply:
I hope that everyone will switch to #26172 and I can close this one. There are no accidentally non-static methods in that PR, and it is a lot cleaner overall.